### PR TITLE
Proportional shapes

### LINKED ISSUE #124: Proportional shapes
Holding control should allow creation of circles instead of ellipses and squares instead of rectangles

### PR DESCRIPTION
**Pull request by sei0o**
Fixes #124 when closed. Currently this provides 1:1/3:2/golden ratio-based constraints for lines, rectangles and ellipses.

Some problems:

- I'm not sure the current way to pass `ratio` to Builders is readable... it could be a parameter for `Shaper::handle_event`, or could be put in `PenEvent`?
- How Foci ellipses and bezier curves should be constrained?

### CODE REVIEW

**Comment by flxzt** [comment]
awesome, thank you for working on this!

* I actually think it is already fine this way being a field in the shaper? We can set the constraint from the UI before we start a shape as you already implemented, or we could listen for the ctrl shortcut key coming with the pen down event to override the setting. I think this would require refactoring the constraint type a little bit to allow an override. Maybe make it a struct with bools to enable multiple constraints in the UI with checkboxes, and one `enabled` to disable the constraints completely when we don't override it. We could then take the struct and override the builders constraints when we press ctrl. Of course we would need then match the constraint that fits the best for the current input in each builder, but this could be done by calculating the distance of the "raw" position to the position when it would be constrained and comparing between them.  
Another thought: There are downsides to passing the constraints at the start of the builder, we can't really change them while building as the user would expect. I think the best solution would be to pass the constraints into the builder in handle_event() so we can decide for every input event.
* I don't know either. Looking at other apps, for example inkscape: it constraints the input for the individual control points, I guess that is a valid way to do it. I think constraints don't really make have a meaning in every builder type, we could also decide that it simply does not have an effect there.  
* Another other thing I noticed: We should definitely always allow horizontal and vertical lines for the line and curve builders. Maybe check with an angle relative to the x-axis ? (check the `angle_ahead()` method for this).

There are a few other things like the drawn indicator for the current input while building the shapes, or the setting location in the UI, but I think that is easily fixed and can be adressed later :)

Oh and I wanted to move away from comborows, because I could not find a way to translate the enum nicks. Check out how it is done in the brushpage, the `texturedstyle_distribution_row` uses a StringList, and I am matching with their indices. This would also make the shaperconstraintratio type unnecessary. Or do you know a way to translate them?

**Comment by sei0o** [comment]
Thanks for the detailed review! 

> Or do you know a way to translate them?

I have no idea \(ツ)/. Other gtk4-rs-based apps such as ([Spot](https://github.com/xou816/spot)) are taking the `GtkStringList` approach as current rnote does.

**Comment by sei0o** [comment]
Now it should enable/disable constraints when `Ctrl` is pressed. They can draw horizontal/vertical lines with newly introduced `Horizontal`/`Vertical` constraints (it might be better to merge these two on user interface) for this. I'm going to support constraints for foci ellipses and bezier curves in days.

**Comment by sei0o** [comment]
It didn't take as much modification as expected. For foci ellipses, I simply put constrains on two foci and it seems fine. For bezier curves, I tried putting constraints between a point and its previous point (e.g. `cp1` and `cp2` on cubic bezier, `cp` and `end` on quadratic bezier), but I'm not confident this behavior is suitable enough since I don't use bezier curves very often.

**Comment by sei0o** [comment]
> There are a few other things like the drawn indicator for the current input while building the shapes, or the setting location in the UI

Could you point them out in detail? :)

File: rnote-engine/src/pens/shaper.rs [hunk 1]
```diff
@@ -55,6 +56,8 @@ pub struct Shaper {
     #[serde(rename = "rough_options")]
     pub rough_options: RoughOptions,
 
+    #[serde(skip)]
+    pub constraint: Constraint,
```
Comment (flxzt) [inline rnote-engine/src/pens/shaper.rs:60]: when Serialize and Deserialize is derived, this field does not need to be skipped. Also singular -> plural

**Review by flxzt — COMMENTED** [review]
If we save and load the constraints with serde, the UI would need to be updated. Check the `refresh-ui-from-engine` action in `appwindow/appwindowactions.rs`, it is called after the engine settings are loaded on start up

**Comment by sei0o** [comment]
> If we save and load the constraints in / from the settings, the UI would need to be updated. Check the refresh-ui-from-engine action in appwindow/appwindowactions.rs, it is called after the engine settings are loaded on start up.

I don't still fully understand rnote's state management, but what about binding GtkSwitches and the gsettings directly in `appsettings.rs`?

File: rnote-compose/src/builders/ellipsebuilder.rs [hunk 2]
```diff
@@ -17,19 +17,24 @@ pub struct EllipseBuilder {
     pub start: na::Vector2<f64>,
     /// the current position
     pub current: na::Vector2<f64>,
+
+    constraint: Constraint,
```
[the remark this answers is not in the document]
Comment (flxzt) [reply rnote-compose/src/builders/ellipsebuilder.rs]: yeah. Do you mind if I add a commit to the branch? I think I have a fix for it. I could also do the wiring to the settings

[the remark this answers is not in the document]
Comment (sei0o) [reply rnote-compose/src/builders/ellipsebuilder.rs]: Of course you can. With regard to the settings, too.

**Comment by flxzt** [comment]
Did some additional polishing to make it ready for merging: moved the constraint applying location in the rect and ellipse builder to fix the wrong red circle indicator. Fixed the golden ratio constraint ( an .abs() was missing ), converted hashmap to hashset ( we don't need the boolean value, if the constraint is in the hash set, it is activated, if not then not). Made the vertical and horizontal constraints internal to the line and curve builders

**Comment by flxzt** [comment]
I think at this state it could be merged :) Another idea I had: the feature could be expanded to allow for custom constraints ( not sure if it is worth the effort ). Your call if you want it to be merged or fleshed out a little more. Also I can take care of the rebasing

**Comment by sei0o** [comment]
Whoa, so quick! I think it's okay not to have custom constraints for now. Could you rebase and merge this then?

**Comment by flxzt** [comment]
Done 🎉

**Comment by sei0o** [comment]
Alright, thanks for the patient review!

**Comment by flxzt** [comment]
anytime :+1:  and thank you for the work!

**Pull request merged**